hydroflow/tests/groupby.rs

```diff
@@ -104,47 +186,45 @@ fn groupby_monotonic_surface() {
     input.give(Iter(BATCH_B.iter().cloned()));
     input.flush();
     hf.tick();
-    assert_eq!(&["mingwei", "justin"], &**output.borrow());
+    assert_eq!(&["sales", "accounting"], &**output.borrow());
 
     input.give(Iter(BATCH_C.iter().cloned()));
     input.flush();
     hf.tick();
-    assert_eq!(&["mingwei", "justin", "mae"], &**output.borrow());
+    assert_eq!(&["sales", "accounting", "engineering"], &**output.borrow());
 }
 
-/// Non-monotonic barrier. Find the median name.
+/// Non-monotonic barrier. Per epoch, for each department find the highest paid employee.
 /// Takes in BATCH_A in the first epoch, then BATCH_B *and* BATCH_C in the second epoch.
+/// SQL (per batch): SELECT department, name, salary FROM employees WHERE salary = MAX(salary) GROUP BY department
 #[test]
 fn groupby_nonmon_surface() {
     use hydroflow::builder::prelude::*;
 
     let mut hf_builder = HydroflowBuilder::new();
-    let (input, source_recv) =
-        hf_builder.add_channel_input::<_, _, VecHandoff<&'static str>>("source");
+    let (input, source_recv) = hf_builder.add_channel_input::<_, _, VecHandoff<Employee>>("source");
 
-    let (a_send, a_recv) = hf_builder.make_edge::<_, VecHandoff<&'static str>, _>("names a-m");
-    let (z_send, z_recv) = hf_builder.make_edge::<_, VecHandoff<&'static str>, _>("names n-z");
+    let (a_send, a_recv) = hf_builder.make_edge::<_, VecHandoff<Employee>, _>("names A-M");
+    let (z_send, z_recv) = hf_builder.make_edge::<_, VecHandoff<Employee>, _>("names N-Z");
     let (stratum_boundary_send, stratum_boundary_recv) =
-        hf_builder.make_edge::<_, VecHandoff<&'static str>, _>("names n-z");
+        hf_builder.make_edge::<_, VecHandoff<Employee>, _>("names N-Z");
 
     // Make output first, to mess with scheduler order.
-    let output = <Rc<Cell<Option<(usize, &'static str)>>>>::default();
+    let output = <Rc<RefCell<BinaryHeap<_>>>>::default();
     let output_ref = output.clone();
     hf_builder.add_subgraph_stratified(
         "find median",
         1,
         stratum_boundary_recv
-            .map(|mut buffer| {
-                let batch_size = buffer.len();
-                let median = *buffer
-                    .make_contiguous()
-                    .select_nth_unstable(batch_size / 2)
-                    .1;
-                (batch_size, median)
+            .flat_map(|batch| {
+                batch
+                    .into_iter()
+                    .into_grouping_map_by(|employee| employee.department)
+                    .max_by_key(|_department, employee| employee.salary)
             })
             .pull_to_push()
-            .map(Some)
-            .for_each(move |val| output_ref.set(val)),
+            .map(|(department, employee)| (department, employee.name, employee.salary))
+            .for_each(move |val| output_ref.borrow_mut().push(val)),
     );
```

Comment:
Yeah this can be changed from
```rust
        stratum_boundary_recv
            .flat_map(|batch| {
                batch
                    .into_iter()
                    .into_grouping_map_by(|employee| employee.department)
                    .max_by_key(|_department, employee| employee.salary)
            })
            .pull_to_push()
            .map(|(department, employee)| (department, employee.name, employee.salary))
            .for_each(move |val| output_ref.borrow_mut().push(val)),
```
to
```rust
        stratum_boundary_recv
            .flat_map(|batch| {
                batch
                    .into_iter()
                    .map(|employee| (employee.department, (employee.name, employee.salary)))
                    .into_grouping_map()
                    .max_by_key(|_department, (_name, salary)| *salary)
            })
            .pull_to_push()
            .map(|(department, (name, salary))| (department, name, salary))
            .for_each(move |val| output_ref.borrow_mut().push(val)),
```